src/Traits/Criteriable.php

```diff
@@ -0,0 +1,269 @@
+<?php
+
+namespace Rinvex\Repository\Traits;
+
+use Closure;
+use Rinvex\Repository\Contracts\CriterionContract;
+use Rinvex\Repository\Contracts\RepositoryContract;
+
+trait Criteriable
+{
+    /**
+     * List of repository criteria.
+     *
+     * @var array
+     */
+    protected $criteria = [];
+
+    /**
+     * List of default repository criteria.
+     *
+     * @var array
+     */
+    protected $defaultCriteria = [];
+
+    /**
+     * Skip criteria flag.
+     * If setted to true criteria will not be apply to the query.
+     *
+     * @var bool
+     */
+    protected $skipCriteria = false;
+
+    /**
+     * Skip default criteria flag.
+     * If setted to true default criteria will not be added to the criteria list.
+     *
+     * @var bool
+     */
+    protected $skipDefaultCriteria = false;
+
+    /**
+     * Return name for the criterion.
+     * If as criterion in parameter passed string we assume that is criterion class name.
+     *
+     * @param CriterionContract|Closure|string $criteria
+     *
+     * @return string
+     */
+    public function getCriterionName($criteria)
+    {
+        if ($criteria instanceof Closure) {
+            return spl_object_hash($criteria);
+        }
+
+        return is_object($criteria) ? get_class($criteria) : $criteria;
+    }
+
+    /**
+     * Add criterion to the specific list.
+     * low-level implementation of adding criterion to the list.
+     *
+     * @param Closure|CriterionContract $criterion
+     * @param string                    $list
+     *
+     * @return $this
+     */
+    protected function addCriterion($criterion, $list)
+    {
+        if (property_exists($this, $list) &&
+            ($criterion instanceof Closure ||
+                $criterion instanceof CriterionContract ||
+                (is_string($criterion) &&
+                    class_exists($criterion) &&
+                    ($criterion = new $criterion()) instanceof CriterionContract
```

Comment:
Might be fatal error here if class constructor have required arguments.
I suggest use ReflectionClass.